### DISCUSSION TITLE
Nuxt 3 Layer: How to customize Layer pages `pageMeta` on consumer side (Nuxt)

### DISCUSSION THREAD

**Discussion by akasection** [nuxt/nuxt#20346] [Questions]
So I have a case in which I have a nuxt layer providing a set of components and two pages without meta. And then on the consumer side (Nuxt), I have an auth middleware and admin layout. 

The question is, how to set/implement the meta (middleware and layout) for the layer pages from the consumer side? I tried using `useAppConfig` but unfortunately, `definePageMeta` is compiler macro, so the `appConfig` is not defined at runtime.

To help illustrating on what I am trying to achieve, here's the picture:

![nuxt_Layer](https://github.com/nuxt/nuxt/assets/1237295/67a2f4d2-1e2e-4e10-bba5-1f6839270e90)

**UPDATE**
We have a workaround for this case by using modules and surgically modifying the metadata onto pages. Not the best but currently works. _Still_, I need more opinion, recommendation or best practice come from Nuxt maintainer since IMO this is pretty dirty and brutal.

The sandbox example: https://codesandbox.io/p/sandbox/nuxt-layer-pagemeta-example-7wws6l

**Comment by Koaieus** [comment]
I have a similar question, related to extending `PageMeta` based on which layers an app is extending.
My idea was that an app that extends the `auth` layer will have its `PageMeta` extended with `{admin?: boolean, allowedRoles?: string[]}`

The solution I found told me to extend the module's interface like so:
```ts
declare module '#app' {
  interface PageMeta {
    admin?: boolean;
    roleNeeded?: string[];
  }
}

export {};
```

The only problem I found with this is that my IDE (WebStorm) does fully recognize this extension when I put this code into a `.d.ts` file in one of the Nuxt apps, but it does NOT get imported properly when placing it in a layer's `index.d.ts` or any other `.d.ts` file. My guess is that this has to do with the order of the imports.

If I explicitly import the layer's `.d.ts` file in the app, the extension does work:
```ts
export * from '<layer>/<d.ts file>';
```

**Comment by akasection** [comment]
Bumped as we found a hacky alternative. Updated on the first post.
